File: Assets/Resources/Scripts/Manager/GenerateManager.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Networking;
using UnityEngine.UI;
using XLua;
/// <summary>
/// 生成图像的代码s
/// </summary>

[CSharpCallLua]
public delegate void LifeCycle(LuaTable luaTable);
public class LuaGenerateManager {
    public LifeCycle start;
    public LifeCycle awake;
    public LifeCycle send;
}
public class GenerateManager : BaseMonoBehaviour {
    //Lua的GenerateManager中的Table
    private LuaGenerateManager luaGenerateManager;
    protected LuaTable luaTable;
    private Image image;
    //延迟发送请求的时间
    private float DELAY_POST_TIME = 4f;
    protected override void FetchComponent() {
        image = this.GetComponent<Image>();
    }

    protected override void GetResources() {//在Awake里执行
        //调lua文件
        XLuaManager.Instance.DoString("require('generate_manager')");
        //得到table
        luaTable = XLuaManager.Instance.Global;
        luaGenerateManager = XLuaManager.Instance.Global.Get<LuaGenerateManager>("generate_manager");
        //把该对象传给Lua
        luaTable.Set<string, GameObject>("obj",this.gameObject);
        //因为lua中没有invoke
        this.Invoke("Send", DELAY_POST_TIME);
    }

    protected override void Awake() {
        base.Awake();
        luaGenerateManager.awake(luaTable);
    }

    protected override void Start() {
        base.Start();
        luaGenerateManager.start(luaTable);
    }

    protected override void InitZoneLayout() {
        
    }

    public void Send() {
        //调lua中的协程
        luaGenerateManager.send(luaTable);
    }
    
    /// <summary>
    /// 发送请求
    /// </summary>
    /// <returns></returns>
    // IEnumerator  SendWebRequest() {
    //     while (true) {
    //         // Debug.Log("发送请求" );
    //         WWWForm form = new WWWForm();
    //         
    //         //mask图像
    //         Texture2D texture = MaskDrawManager.instance.GetMaskTexture();
    //         
    //         
    //         //当前选中的角色
    //         int label_class = CharacterChooseManager.instance.GetCurrentChooseCharacterID();
    //         //当前选择的模型
    //         string model_id = SceneDataManager.GetInstance().GetModelID();
    //         
    //         form.AddBinaryData("label_img",texture.EncodeToPNG());
    //
    //         form.AddField("model_id", model_id);
    //         form.AddField("label_class", label_class);
    //         UnityWebRequest www = UnityWebRequest.Post(Config.POST_HTTP_URL, form);
    //        
    //         yield return www.SendWebRequest();
    //         if (www.isDone) {
    //             if (www.result == UnityWebRequest.Result.ConnectionError ||
    //                 www.result == UnityWebRequest.Result.ProtocolError) {
    //                 Debug.LogWarning(www.error);
    //             } else {
    //                 string responseText = www.downloadHandler.text;
    //                 //获取图像
    //                 www = UnityWebRequestTexture.GetTexture(Config.DOWNLOAD_HTTP_URL+responseText);
    //                 yield return www.SendWebRequest();
    //                 if (www.isDone) {
    //                     if (www.result == UnityWebRequest.Result.ConnectionError ||
    //                         www.result == UnityWebRequest.Result.ProtocolError) {
    //                         Debug.LogWarning(www.error);
    //                     } else {
    //                         texture = ((DownloadHandlerTexture)www.downloadHandler).texture;
    //                         Sprite sprite = Sprite.Create(texture, new Rect( 0 ,0, texture.width, texture.height), Vector2.zero);
    //                         
    //                         image.sprite = sprite;
    //                     }
    //                 }
    //             }
    //         }
    //     }
    // }

   
}

File: Assets/Resources/Lua/generate_manager.lua

generate_manager = {}

local util = require 'xlua.util'
-- 生命周期awke
function generate_manager:awake()
    print("lua:awake")
    generate_manager.image = obj.transform:GetComponent("Image")
    generate_manager.CSharpGenerateManager = obj:GetComponent("GenerateManager")
end
--生命周期start
function generate_manager:start()
    print("luar:start")
end

--开启协程
function generate_manager:send()
    generate_manager.CSharpGenerateManager:StartCoroutine(generate_manager.send_web_request)
end

--协程
generate_manager.send_web_request = util.cs_generator(function()
    print("luar:StartCoroutine ")
    while true do
        local form = CS.UnityEngine.WWWForm()
        -- print(CS.CharacterChooseManager.instance)
        -- mask图像 
        local texture = CS.MaskDrawManager.instance:GetMaskTexture()
        --当前选中的角色
        local label_class = CS.CharacterChooseManager.instance:GetCurrentChooseCharacterID();
        -- 当前选择的模型
        local model_id = CS.SceneDataManager.GetInstance():GetModelID()
        -- mask图片
        -- local label_img = texture.EncodeToPNG()
        -- print(CS.UnityEngine.ImageConversion)
        local label_img = CS.UnityEngine.ImageConversion.EncodeToPNG(texture)
        -- 参数
        form:AddBinaryData("label_img",label_img)
        form:AddField("model_id", model_id)
        form:AddField("label_class", label_class)

        local www =  CS.UnityEngine.Networking.UnityWebRequest.Post(CS.Config.POST_HTTP_URL, form);
        --print(CS.Config.POST_HTTP_URL)
        -- 发送请求
        coroutine.yield(www:SendWebRequest())
        if www.isDone then
            if www.result == CS.UnityEngine.Networking.UnityWebRequest.Result.ConnectionError and  www.result == CS.UnityEngine.Networking.UnityWebRequest.Result.ProtocolError then
                CS.UnityEngine.Debug.LogWarning(www.error)
            else
                -- 得到回复
                local responseText = www.downloadHandler.text
                -- 再发送请求得到图片
                www = CS.UnityEngine.Networking.UnityWebRequestTexture.GetTexture(CS.Config.DOWNLOAD_HTTP_URL..responseText)
                coroutine.yield(www:SendWebRequest())
                if www.isDone then
                    if www.result == CS.UnityEngine.Networking.UnityWebRequest.Result.ConnectionError and  www.result == CS.UnityEngine.Networking.UnityWebRequest.Result.ProtocolError then
                        CS.UnityEngine.Debug.LogWarning(www.error)
                    else
                        texture = www.downloadHandler.texture
                        local rect = CS.UnityEngine.Rect(0,0,texture.width,texture.height)
                        local sprite = CS.UnityEngine.Sprite.Create(texture,rect,CS.UnityEngine.Vector2.zero)
                        generate_manager.image.sprite = sprite
                    end
                end
               

            end
            
        end
        
    end
end)

